Sources/cte/Parser.swift

```diff
@@ -159,6 +159,66 @@ struct Parser {
             let stmtIf = AstNode.If(condition: cond, thenStmt: body, elseStmt: elseStmt)
             return AstNode(stmtIf, tokens: [ifToken, elseToken])
 
+        case .keywordFor:
+            let forToken = advance()
+
+            // Check for infinite loop
+            guard lexer.peek()?.kind != .lbrace else {
+                let body = expression()
+                let stmtFor = AstNode.For(initialiser: nil, condition: nil, step: nil, body: body)
+                return AstNode(stmtFor, tokens: [forToken])
+            }
+
+            var exprs: [AstNode] =  []
+            while exprs.count <= 3 {
+                let token = lexer.peek()?.kind
+                guard token != .semicolon else {
+                    let semicolon = advance()
+                    exprs.append(AstNode(AstNode.Empty(), tokens: [semicolon]))
+                    continue
+                }
+
+                guard token != .lbrace else {
+                    // Check if we have 2 cases already. If so, that means the
+                    // 3rd case is empty. If not, we unexpectedly hit an `{`
+                    if exprs.count == 2 {
```

Comment:
My only other concern is that this will likely fail unintuitively for something like `for i := 0; j := 1; i < 5 && j < 5; i += 1; j += 1`
I know that's a very odd example but I think just changing to consuming until peeking returns a `lparen` then afterwards switching on the count will result in cleaner code here.

```swift
var exprs: [AstNode] = []
while lexer.peek()?.kind != .lparen {
    let expr = expression()
    exprs.append(expr)
}

let body = expression()

var initialiser: AstNode?
var condition: AstNode?
var step: AstNode?

switch exprs.count {
case 0:
    break
case 1:
    condition = exprs[0]
case 3:
    initialiser = exprs[0]
    condition = exprs[1]
    step = exprs[2]
default:
    reportError("`for` statements require 0, 1 or 3 statements", at: forToken)
    return AstNode.invalid(with: [forToken])
}

let för = AstNode.For(initialiser: initialiser, condition: condition, step: step, body: body)
return AstNode(för, tokens: [forToken])
```